File: AndroidStudy/trail_running/src/main/java/com/example/trail_running/MainActivity.java
package com.example.trail_running;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import android.app.Activity;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.Window;

import com.google.android.material.tabs.TabLayout;

public class MainActivity extends Activity {
    private TabLayout tab_lepao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        this.requestWindowFeature(Window.FEATURE_NO_TITLE);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViews();
        TabLayout.Tab dongtai = tab_lepao.newTab();//内部类
        dongtai.setIcon(R.mipmap.dongtai);
        dongtai.setText("动态");

        TabLayout.Tab xiaoyuan = tab_lepao.newTab();
        xiaoyuan.setIcon(R.mipmap.xiaoyuan);
        xiaoyuan.setText("校园");

        TabLayout.Tab lepao = tab_lepao.newTab();
        lepao.setIcon(R.mipmap.lepao);

        lepao.setText("乐跑");

        TabLayout.Tab wode = tab_lepao.newTab();
        wode.setIcon(R.mipmap.mine);
        wode.setText("我的");

        int gray = ContextCompat.getColor(this, R.color.gray);
        int green = ContextCompat.getColor(this, R.color.green);
        tab_lepao.setTabTextColors(gray, green);

        tab_lepao.addTab(dongtai);
        tab_lepao.addTab(xiaoyuan);
        tab_lepao.addTab(lepao);
        tab_lepao.addTab(wode);

        tab_lepao.getTabAt(2).select();

        tab_lepao.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                int position = tab.getPosition();
                switch (position){
                    case 0:
                        dongtai.setIcon(R.mipmap.dongtai_green);
                        xiaoyuan.setIcon(R.mipmap.xiaoyuan);
                        lepao.setIcon(R.mipmap.lepao);
                        wode.setIcon(R.mipmap.mine);
                        break;
                    case 1:
                        dongtai.setIcon(R.mipmap.dongtai);
                        xiaoyuan.setIcon(R.mipmap.xiaoyuan_green);
                        lepao.setIcon(R.mipmap.lepao);
                        wode.setIcon(R.mipmap.mine);
                        break;
                    case 2:
                        dongtai.setIcon(R.mipmap.dongtai);
                        xiaoyuan.setIcon(R.mipmap.xiaoyuan);
                        lepao.setIcon(R.mipmap.lepao_green);
                        wode.setIcon(R.mipmap.mine);
                        break;
                    case 3:
                        dongtai.setIcon(R.mipmap.dongtai);
                        xiaoyuan.setIcon(R.mipmap.xiaoyuan);
                        lepao.setIcon(R.mipmap.lepao);
                        wode.setIcon(R.mipmap.mine_green);
                        break;
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });

    }

    private void findViews() {
        tab_lepao = findViewById(R.id.tab_lepao);
    }
}

File: KotlinStudy/app/src/main/java/com/example/kotlinstudy/ui/activity/MainActivity.kt
package com.example.kotlinstudy.ui.activity

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.example.kotlinstudy.R
import com.example.kotlinstudy.databinding.ActivityMainBinding
import com.example.kotlinstudy.utils.db.entity.UserInfo
import com.example.kotlinstudy.ui.fragment.HomeFragment
import com.example.kotlinstudy.ui.fragment.MineFragment
import com.example.kotlinstudy.viewmodel.MyViewModel
import com.google.android.material.tabs.TabLayout

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"

    private lateinit var binding : ActivityMainBinding;
    private  var thisFragemntId : Int? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if(savedInstanceState != null){
            thisFragemntId = savedInstanceState.getInt("thisFragmentId")
        }
        if(thisFragemntId == null)thisFragemntId = 0

        initTab()
        initViewModel()
    }

    private fun initViewModel(){
        //获取ViewModel视图对象
        var viewModel : MyViewModel = ViewModelProvider(
            this,
            MyViewModel.MyViewModelFactory(application)
        ).get(MyViewModel::class.java)

        //为ViewModel中获取的LiveData数据设置observe监听，当数据库中数据改变时，自动回调onChanged方法
        viewModel.query().observe(this, object : Observer<List<UserInfo>>{
            override fun onChanged(value: List<UserInfo>) {
                Log.i(TAG, "onChanged: Observer#onChanged回调, List<UserInfo> : " + value)
            }
        })
    }

    private fun initTab(){
        val tab : TabLayout = binding.mainTab;

        tab.addTab(tab.newTab().setText("Login"))
        tab.addTab(tab.newTab().setText("Mine"))

        when(thisFragemntId){
            0 -> loadFragment(HomeFragment())
            else -> loadFragment(MineFragment())
        }


        tab.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener{
            override fun onTabSelected(tab: TabLayout.Tab?) {
                thisFragemntId = tab?.position!!
                val fragment = when(tab?.position){
                    0 -> HomeFragment()
                    1 -> MineFragment()
                    else -> null
                }
                //使用let安全调用fragment作为参数的方法
                fragment?.let { loadFragment(it) }
            }

            override fun onTabUnselected(tab: TabLayout.Tab?) {
                println("Not yet implemented")
            }

            override fun onTabReselected(tab: TabLayout.Tab?) {
                println("Not yet implemented")
            }

        })
    }

    private fun loadFragment(fragment: Fragment){
        supportFragmentManager.beginTransaction()
            .replace(R.id.main_fcv, fragment)
            .commit()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("thisFragmetId", thisFragemntId!!)
    }



}